src/AbstractAppliance.js

```diff
@@ -0,0 +1,112 @@
+ *
+ * AbstractAppliance methods prefixed with `_` require implementation.
+ * Other methods, aside from the constructor, should not be overridden.
+ *
+ * For more information about the TV Kitchen architecture visit:
+ * https://github.com/tvkitchen/tv-kitchen/blob/master/docs/ARCHITECTURE.md
+ */
+class AbstractAppliance {
+	settings = {}
+
+	emitter = new EventEmitter()
+
+	payloadBuffer = new PayloadBuffer()
+
+	constructor(overrideSettings) {
+		Object.assign(
+			this.settings,
+			this.getDefaultSettings(),
+			overrideSettings,
+		)
+	}
+
+	/*******************************************
+	 ************ ABSTRACT METHODS *************
+	 *******************************************/
+
+	/**
+	 * Getter for the default settings of the appliance.
+	 * Used by the core constructor and also serves as self-documentation.
+	 *
+	 * OVERRIDE WHEN EXTENDING
+	 *
+	 * @return {Object} The default settings of the appliance
+	 */
+	getDefaultSettings = () => throw new NotImplementedError('getDefaultSettings')
+
+	/**
+	 * Getter for the list of data types accepted by the appliance.
+	 *
+	 * OVERRIDE WHEN EXTENDING
+	 *
+	 * @return {Array[String]} The list of data types accepted by the appliance.
+	 */
+	getInputTypes = () => throw new NotImplementedError('getInputTypes')
+
+	/**
+	 * Getter for the list of data types produced by the appliance.
+	 *
+	 * OVERRIDE WHEN EXTENDING
+	 *
+	 * @return {Array[String]} The list of data types produced by the appliance.
+	 */
+	getOutputTypes = () => throw new NotImplementedError('getOutputTypes')
+
+	/**
+	 * Asserts that a given payload is actually a payload.
+   *
+   * OVERRIDE WHEN EXTENDING
+   *
+	 * @param  {Payload} payload The payload that we want to validate.
+	 * @return {Boolean}         The result of the validation check.
+	 */
+	// eslint-disable-next-line no-unused-vars
+	isValidPayload = (payload) => throw new NotImplementedError('isValidPayload')
+
+	/**
+	 * Invokes the appliance on any unprocessed data in the appliance buffer.
+	 *
+	 * @return {Boolean}         Whether or not the appliance had enough data to work with.
+	 */
+	 // eslint-disable-next-line no-unused-vars
+	invoke = () => throw new NotImplementedError('process')
+
+	/*******************************************
+	 *********** IMPLEMENTED METHODS ***********
+	 *******************************************/
+
+	/**
+	 * Called by a third party to pass data into the appliance.
+	 * If the payload is valid, it is added to the buffer and the appliance is invoked.
+	 *
+	 * @throws
+	 * @param  {[type]} payload [description]
+	 * @return {[type]}         [description]
+	 */
+	ingestPayload = (payload) => {
+		this.emitter.emit('input', { payload })
+		assert(isPayloadInstance(payload))
+		assert(this.isValidPayload(payload))
+		this.payloadBuffer.add(payload)
+		return this.invoke()
+	}
+
+	on = (eventType, callback) => emitter.on(eventType, listener)
+
+	emitPayload = (payload) => emitter.emit('output', { payload })
```

Comment:
This will need to be called by the appliance itself when data is ready for emission.  Since we don't know the nature of the appliance we should not "force" that to happen.